cdap-security-authz-dataset/src/main/java/co.cask.cdap.security.authz/ACLDataset.java

```diff
@@ -0,0 +1,136 @@
+import co.cask.cdap.api.dataset.table.Table;
+import co.cask.cdap.data2.dataset2.lib.table.MDSKey;
+import co.cask.cdap.proto.Id;
+import co.cask.cdap.proto.id.EntityId;
+
+import java.util.HashSet;
+import java.util.Set;
+
+/**
+ *
+ */
+class ACLDataset extends AbstractDataset {
+
+  public static final String NAME = "acls";
+  public static final Id.DatasetInstance ID = Id.DatasetInstance.from(Id.Namespace.SYSTEM, NAME);
+
+  private static final byte[] VALUE_COLUMN = new byte[0];
+
+  private final Table table;
+
+  ACLDataset(Table table) {
+    super(NAME, table);
+    this.table = table;
+  }
+
+  /**
+   * @param entity the entity
+   * @param user the user
+   * @return the set of actions allowed for the user on the entity
+   */
+  public Set<Action> search(EntityId entity, String user) {
+    Set<Action> result = new HashSet<>();
+
+    MDSKey mdsKey = getKey(entity, user);
+    byte[] startKey = mdsKey.getKey();
+    byte[] stopKey = Bytes.stopKeyForPrefix(startKey);
+    Scanner scan = table.scan(startKey, stopKey);
+
+    try {
+      Row next;
+      while ((next = scan.next()) != null) {
+        byte[] value = next.get(VALUE_COLUMN);
+        if (value == null) {
+          continue;
+        }
+        result.add(Action.valueOf(Bytes.toString(value)));
+      }
+    } finally {
+      scan.close();
+    }
+
+    return result;
+  }
+
+  public void add(EntityId entity, String user, Action action) {
+    table.put(getKey(entity, user, action).getKey(), VALUE_COLUMN, Bytes.toBytes(action.name()));
+  }
+
+  public void remove(EntityId entity, String user, Action action) {
+    table.delete(getKey(entity, user, action).getKey());
+  }
+
+  public void remove(EntityId entity, String user) {
+    MDSKey mdsKey = getKey(entity, user);
+    byte[] startKey = mdsKey.getKey();
+    byte[] stopKey = Bytes.stopKeyForPrefix(startKey);
+    Scanner scan = table.scan(startKey, stopKey);
+
+    try {
+      Row next;
+      while ((next = scan.next()) != null) {
+        table.delete(next.getRow());
+      }
+    } finally {
+      scan.close();
+    }
+  }
+
+  public void remove(EntityId entity) {
+    MDSKey mdsKey = getKey(entity);
+    byte[] startKey = mdsKey.getKey();
+    byte[] stopKey = Bytes.stopKeyForPrefix(startKey);
+    Scanner scan = table.scan(startKey, stopKey);
+
+    try {
+      Row next;
+      while ((next = scan.next()) != null) {
+        table.delete(next.getRow());
+      }
+    } finally {
+      scan.close();
+    }
+  }
+
+  private MDSKey getKey(EntityId entity) {
+    MDSKey.Builder builder = new MDSKey.Builder();
+    builder.add(entity.toString());
+    return builder.build();
+  }
+
+  private MDSKey getKey(EntityId entity, String user) {
+    MDSKey.Builder builder = new MDSKey.Builder();
+    builder.add(entity.toString());
+    builder.add(user);
+    return builder.build();
+  }
+
+  private MDSKey getKey(EntityId entity, String user, Action action) {
+    MDSKey.Builder builder = new MDSKey.Builder();
```

Comment:
I don't see much use in trying to reduce code duplication here, as it will make it harder to read and we don't gain much.